pkg/store/proxy.go

```diff
@@ -280,10 +306,22 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 		return mergedSet.Err()
 	})
 
+	// All code above is fast and light on memory.
+	// The real work is in the loop that follows.
+	// Before we do anything expense, check we should even start.
+	if err := checkMem(); err != nil {
+		return err
+	}
+
 	for resp := range respRecv {
 		if err := srv.Send(resp); err != nil {
 			return status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
 		}
+
+		// bail if exceeded mem.
+		if err := checkMem(); err != nil {
```

Comment:
I think we might to that on receive rather.. just thinking loudly

Reply:
is this the wrong place to return an error?

Reply:
Return error, sure, but checking - it sounds more natural to check it when you receive the data

Reply:
Where would that be? Sorry I'm not as familiar with codebase yet :) 

Do you mean inside `srv.Send(resp)` perhaps? Does it even matter? To me this does look like receiver of sorts, since this loop reads out from `respRecv` channel and the var is called response?